packages/fetchai/connections/p2p_libp2p/connection.py

```diff
@@ -129,26 +131,121 @@ def _golang_module_run(
+        self._wait_status: Optional[asyncio.Future] = None
 
     async def connect(self) -> bool:
         """Connect to node with pipe."""
         return await self.pipe.connect()
 
     async def send_envelope(self, envelope: Envelope) -> None:
         """Send envelope to node."""
-        await self._write(envelope.encode())
+        self._wait_status = asyncio.Future()
+        buf = self.make_acn_envelope_message(envelope)
+        await self._write(buf)
+        try:
+            status = await self.wait_for_status()
+            if status.code != Status.SUCCESS:  # type: ignore  # pylint: disable=no-member
+                raise Exception(
+                    f"failed to send envelope. got error confirmation: {status}"
+                )
+        except asyncio.TimeoutError:
+            if not self._wait_status.done():  # pragma: nocover
+                self._wait_status.set_exception(Exception("Timeout"))
+            await asyncio.sleep(0)
+            raise Exception("acn status await timeout!")
+        finally:
+            self._wait_status = None
+
+    async def wait_for_status(self) -> Any:
+        """Get status."""
+        if self._wait_status is None:  # pragma: nocover
+            raise ValueError("value failed!")
+        return await asyncio.wait_for(self._wait_status, timeout=self.ACN_ACK_TIMEOUT)
+
+    def make_acn_envelope_message(self, envelope: Envelope) -> bytes:
+        """Make acn message with envelope in."""
+        aea_envelope = AeaEnvelope()
+        aea_envelope.record.CopyFrom(  # pylint: disable=no-member
+            self.make_agent_record()
+        )
+        aea_envelope.envel = envelope.encode()
+        msg = AcnMessage()
+        msg.version = ACN_CURRENT_VERSION
+        msg.aea_envelope.CopyFrom(aea_envelope)  # pylint: disable=no-member
+        buf = msg.SerializeToString()
+        return buf
+
+    def make_agent_record(self) -> AgentRecordPb:  # type: ignore
+        """Make acn agent record."""
+        record = AgentRecordPb()
+        record.address = self.agent_record.address
+        record.public_key = self.agent_record.public_key
+        record.peer_public_key = self.agent_record.representative_public_key
+        record.signature = self.agent_record.signature
+        record.service_id = POR_DEFAULT_SERVICE_ID
+        record.ledger_id = self.agent_record.ledger_id
+        return record
 
     async def read_envelope(self) -> Optional[Envelope]:
         """Read envelope from the node."""
-        data = await self._read()
+        while True:
+            data = await self._read()
 
-        if not data:
-            return None
+            if not data:
+                return None
 
-        return Envelope.decode(data)
+            try:
+                msg = AcnMessage()
+                msg.ParseFromString(data)
+            except Exception as e:
+                await self.write_acn_status_error(f"Failed to parse acn message {e}")
```

Comment:
here the error should be error_serialization

Reply:
done